linker/linker_soinfo.cpp

```diff
@@ -745,6 +746,27 @@ void soinfo::generate_handle() {
   g_soinfo_handles_map[handle_] = this;
 }
 
+ElfW(Addr) soinfo::memory_vaddr(ElfW(Addr) file_vaddr) const {
+  ElfW(Addr) res = file_vaddr + load_bias;
+  size_t size = rand_addr_segments.size();
+  if (size == 0)
+    return res;
+
+  for (int l = 0, h = size-1; l <= h; ) {
+    unsigned i = (l+h)/2;
+    const SegmentInfo &seg_info = rand_addr_segments[i];
+    if (file_vaddr >= seg_info.phdr_addr) {
+      if (file_vaddr < (seg_info.phdr_addr + seg_info.real_size)) {
+        return seg_info.real_addr + (file_vaddr - PAGE_START(seg_info.phdr_addr));
+      } else
+        l = i+1;
+    } else {
+      h = i-1;
+    }
+  }
+  return res;
+}
+
```

Comment:
I checked, looks like you can use anything in libcxx headers, and libc++_static. `std::lower_bound` and `std::binary_search` both look to be ok.